docs/RFCS/20190906_flow_control.md

```diff
@@ -0,0 +1,137 @@
+
+Take the [benchmark result in DM 1.0-GA](https://pingcap.com/docs/v3.0/benchmark/dm-v1.0-ga/#benchmark-result-with-different-pool-size-in-load-unit) as an example, we extract the result in load unit test, use load unit pool size as X-axis, latency and load speed as Y-axis.
+
+![DM benchmark result with different pool size in load unit](../media/rfc-load-benchmark.png)
+
+From the benchmark result we can draw the following conclusions:
+
+- The latency grows exponentially as the concurrency increases.
+- The TPS grows linely as the concurrency increases in a specific range, in this case the liner range is between 0-16.
+- After the TPS reaches the maximum value, it remains the same even the concurrency increases. The maximum value of TPS can be used as the bandwidth.
+
+#### How to detect and measure congestion
+
+Congestion can be exposed in different ways and have different effect on key indicators. We have the following predictive indicators change that may reflect congestion.
+
+- If TPS doesn't increase alongside with the concurrency increase (assuming the dumped data or relay log stream is fast enough), we can believe the congestion is occurring.
+- If concurrency doesn't change, but TPS decreases and latency increases significantly (maybe TPS halves and latency doubles), we can believe the congestion is occurring.
+- If concurrency doesn't change, TPS and latency don't change too much, but data loss rate increases significantly (however in most cases if data loss rate increases, we often find TPS decreases and latency increases), we can believe the congestion is occurring.
+
+The congestion degree can be measured by the three indicators, but the workflow in the real world is often complicated, we won't introduce an estimate model in this proposal, as we only need to known whether the congestion is occurring in our congestion control model.
+
+### Congestion control framework
+
+The congestion control framework is separated into three parts, the congestion detect component, the data flow control component and the concurrency control component.
+
+#### The congestion detect component
+
+The congestion detect component will collect the key indicators in DM-worker as following:
+
+- data loss rate: we will add a sliding window to record the count of downstream resource busy error (such as `tmysql.ErrPDServerTimeout`, `tmysql.ErrTiKVServerTimeout`, `tmysql.ErrTiKVServerBusy` etc) and the count of all dispatched transactions. The data loss rate will be calculated with these two counters.
+- TPS detection: we will add a sliding window to record the count of successfully executed transactions, and calculate TPS with this value.
+- latency: transaction execution latency is already recorded in load and sync unit, we can use it directly.
+
+#### The data flow control component
```

Comment:
I think `the data flow control component` can merge into `The congestion detect component

Reply:
IMO these two components are designed for different purposes
- The congestion detect component collects information for congestion control
- The `data flow control component` takes the input from congestion information and outputs a certain concurrency.
It is easy to implement them as two separated components.